docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,765 @@
+will be picked up by Elasticsearch and the new hosts list will be used.
+
+Note that the file-based discovery plugin augments the unicast hosts list in
+`elasticsearch.yml`: if there are valid unicast host entries in
+`discovery.zen.ping.unicast.hosts` then they will be used in addition to those
+supplied in `unicast_hosts.txt`.
+
+The `discovery.zen.ping.unicast.hosts.resolve_timeout` setting also applies to
+DNS lookups for nodes specified by address via file-based discovery. This is
+specified as a <<time-units, time unit>> and defaults to 5s.
+
+The format of the file is to specify one node entry per line.  Each node entry
+consists of the host (host name or IP address) and an optional transport port
+number.  If the port number is specified, is must come immediately after the
+host (on the same line) separated by a `:`.  If the port number is not
+specified, a default value of 9300 is used.
+
+For example, this is an example of `unicast_hosts.txt` for a cluster with four
+nodes that participate in unicast discovery, some of which are not running on
+the default port:
+
+[source,txt]
+----------------------------------------------------------------
+10.10.10.5
+10.10.10.6:9305
+10.10.10.5:10005
+# an IPv6 address
+[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:9301
+----------------------------------------------------------------
+
+Host names are allowed instead of IP addresses (similar to
+`discovery.zen.ping.unicast.hosts`), and IPv6 addresses must be specified in
+brackets with the port coming after the brackets.
+
+It is also possible to add comments to this file. All comments must appear on
+their lines starting with `#` (i.e. comments cannot start in the middle of a
+line).
+
+[float]
+[[ec2-hosts-provider]]
+===== EC2 hosts provider
+
+The {plugins}/discovery-ec2.html[EC2 discovery plugin] adds a hosts provider
+that uses the https://github.com/aws/aws-sdk-java[AWS API] to find a list of
+seed nodes.
+
+[float]
+[[azure-classic-hosts-provider]]
+===== Azure Classic hosts provider
+
+The {plugins}/discovery-azure-classic.html[Azure Classic discovery plugin] adds
+a hosts provider that uses the Azure Classic API find a list of seed nodes.
+
+[float]
+[[gce-hosts-provider]]
+===== Google Compute Engine hosts provider
+
+The {plugins}/discovery-gce.html[GCE discovery plugin] adds a hosts provider
+that uses the GCE API find a list of seed nodes.
+
+[float]
+==== Discovery settings
+
+Discovery operates in two phases: First, each node probes the addresses of all
+known master-eligible nodes by connecting to each address and attempting to
+identify the node to which it is connected. Secondly it shares with the remote
+node a list of all of its known master-eligible peers and the remote node
+responds with _its_ peers in turn. The node then probes all the new nodes about
+which it just discovered, requests their peers, and so on, until it has
+discovered an elected master node or enough other masterless master-eligible
```

Comment:
Can we separately describe the process for master-eligible and not a master-eligible node? Because only master-eligible node will start the election, non-master eligible node should discover already elected master.

Reply:
Good idea. I pushed e6087e95dde43744ded7263aa609916d368faf4b.